lib/services/unified_resource.go

```diff
@@ -0,0 +1,514 @@
+			if limit > 0 && len(res) >= limit {
+				return false
+			}
+			return true
+		})
+		return nil
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(res) == backend.DefaultRangeLimit {
+		c.log.Warnf("Range query hit backend limit. (this is a bug!) startKey=%q,limit=%d", startKey, backend.DefaultRangeLimit)
+	}
+
+	return res, nil
+}
+
+// GetUnifiedResources returns a list of all resources stored in the current unifiedResourceCollector tree
+func (c *UnifiedResourceCache) GetUnifiedResources(ctx context.Context) ([]types.ResourceWithLabels, error) {
+	result, err := c.getRange(ctx, backend.Key(prefix), backend.RangeEnd(backend.Key(prefix)), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err, "getting unified resource range")
+	}
+
+	resources := make([]types.ResourceWithLabels, 0, len(result))
+	for _, item := range result {
+		resources = append(resources, item.CloneResource())
+	}
+
+	return resources, nil
+}
+
+// ResourceGetter is an interface that provides a way to fetch all the resources
+// that can be stored in the UnifiedResourceCache
+type ResourceGetter interface {
+	NodesGetter
+	DatabaseServersGetter
+	AppServersGetter
+	WindowsDesktopGetter
+	KubernetesServerGetter
+	SAMLIdpServiceProviderGetter
+}
+
+// newWatcher starts and returns a new resource watcher for unified resources.
+func newWatcher(ctx context.Context, resourceCache *UnifiedResourceCache, cfg ResourceWatcherConfig) error {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err, "setting defaults for unified resource watcher config")
+	}
+
+	if _, err := newResourceWatcher(ctx, resourceCache, cfg); err != nil {
+		return trace.Wrap(err, "creating a new unified resource watcher")
+	}
+	return nil
+}
+
+func resourceKey(resource types.Resource) []byte {
+	return backend.Key(prefix, resource.GetName(), resource.GetKind())
+}
+
+func (c *UnifiedResourceCache) getResourcesAndUpdateCurrent(ctx context.Context) error {
+	newNodes, err := c.getNodes(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	newDbs, err := c.getDatabaseServers(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	newKubes, err := c.getKubeServers(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	newApps, err := c.getAppServers(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	newSAMLApps, err := c.getSAMLApps(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	newDesktops, err := c.getDesktops(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return c.read(ctx, func(tree *btree.BTreeG[*item]) error {
+		tree.Clear(false)
+		putResources[types.Server](tree, newNodes)
+		putResources[types.DatabaseServer](tree, newDbs)
+		putResources[types.KubeServer](tree, newKubes)
+		putResources[types.AppServer](tree, newApps)
+		putResources[types.SAMLIdPServiceProvider](tree, newSAMLApps)
+		putResources[types.WindowsDesktop](tree, newDesktops)
+
+		c.stale = false
+		c.defineCollectorAsInitialized()
+		return nil
+	})
```

Comment:
This shouldn't be using `read`.  The `read` function is a helper specifically for falling back to the TTL-cache if the primary is stale/unhealthy.  Initialization is supposed to always insert into the primary `c.tree`.  I'd move this block out of `read` and just hold the lock normally here instead.  As currently implemented, the initial state will end up getting written into the ttl-cache's tree, and basically get 'lost' since we mark the primary as healthy right after insertion.

It looks like current test coverage only starts creating resources after the cache has become healthy, meaning that the test coverage effectively isn't checking if this initialization logic works (since the initial state is always an empty tree), its just checking if event handling works (which is also important).  I'd suggest adding some test coverage that ensures that the cache initialize correctly in the presence of existing resources.